test/jdk/javax/swing/JMenuItem/8031573/bug8031573.java

```diff
@@ -20,45 +20,105 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
+
+import java.awt.BorderLayout;
 import java.awt.FlowLayout;
-import javax.swing.JApplet;
+import java.awt.event.WindowAdapter;
+import java.awt.event.WindowEvent;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import javax.swing.JButton;
 import javax.swing.JCheckBoxMenuItem;
+import javax.swing.JFrame;
 import javax.swing.JMenu;
 import javax.swing.JMenuBar;
+import javax.swing.JMenuItem;
+import javax.swing.JPanel;
+import javax.swing.JTextArea;
+import javax.swing.JTextField;
 import javax.swing.SwingUtilities;
 import javax.swing.UIManager;
+import javax.swing.text.JTextComponent;
 
 /* @test
  * @bug 8031573 8040279 8143064
  * @summary [macosx] Checkmarks of JCheckBoxMenuItems aren't rendered
  *           in high resolution on Retina
- * @author Alexander Scherbatiy
- * @run applet/manual=yesno bug8031573.html
+ * @run main/manual bug8031573
  */
-public class bug8031573 extends JApplet {
 
-    @Override
-    public void init() {
-        try {
+public class bug8031573 {
+
+    private static JFrame frame;
+    private static volatile boolean passed = false;
+    private static final CountDownLatch latch = new CountDownLatch(1);
 
-            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
+    public static final String INSTRUCTIONS = "INSTRUCTIONS:\n\n"
+            + "Verify that high resolution system icons are used JCheckBoxMenuItem on HiDPI displays.\n"
+            + "If the display does not support HiDPI mode press PASS.\n"
+            + "1. Run the test on HiDPI Display.\n"
+            + "2. Press the Menu in the applet.\n"
+            + "3. Check that the icon on the JCheckBoxMenuItem is smooth If so, press PASS, else press FAIL.\n";
 
-            SwingUtilities.invokeAndWait(new Runnable() {
+    public static void main(String args[]) throws Exception {
+        UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
+        SwingUtilities.invokeAndWait(() -> createTestGUI());
 
-                @Override
-                public void run() {
-                    JMenuBar bar = new JMenuBar();
-                    JMenu menu = new JMenu("Menu");
-                    JCheckBoxMenuItem checkBoxMenuItem
-                            = new JCheckBoxMenuItem("JCheckBoxMenuItem");
-                    checkBoxMenuItem.setSelected(true);
-                    menu.add(checkBoxMenuItem);
-                    bar.add(menu);
-                    setJMenuBar(bar);
-                }
-            });
-        } catch (Exception e) {
-            throw new RuntimeException(e);
+        if (!latch.await(5, TimeUnit.MINUTES)) {
+            frame.dispose();
```

Comment:
The `dispose()` method should be called on EDT.

Reply:
If I remove the frame.dispose(), the frame doesn't gets disposed even after timeout.

Reply:
You're right, in my previous comment I forgot about timeout.
Yet it should still be called on EDT.

However, to clean up the code a bit, I propose calling `dispose()` in finally block. Thus there'll be only one place where frame is disposed of instead of being scattered in multiple event handlers:
```
try {
    // Create GUI
    // Check conditions and throw exceptions on failure
} finally {
    SwingUtilities.invokeAndWait(() -> {
        if (frame != null) {
            frame.dispose();
        }
    }
}
```
In this case, frame will be accessed from one thread only, EDT, and does not need to be volatile.

Since the frame is always disposed of, other calls to `frame.dispose()` can be removed.